### PR TITLE
Added support to typescript repos

### PR DESCRIPTION
**Pull request by mpatsute**
This PR is for adding typescript support. The requirement is that we should be able to run "sdocs" command in the repo which has type script (.ts) files. 

Implementation Details: On seeing the file extension ".ts", we invoke the typeDoc parser on this files, which gives us the parsed json. We then invoke the strong-docs rendering engine, to render this parsed json. 

Important Modules: 
lib/tsParser.js
lib/tsConstruct.js
templates/tsConstructs/*.ejs
test/ts-parser.test.js

### CODE REVIEW

File: bin/cli.js [hunk 1]
```diff
@@ -68,7 +70,7 @@ function getAssetData(config) {
  * Preview mode 
  */
 
-if(previewMode) {
+if(previewMode  && type !== 'ts') {
```
Comment (superkhau) [inline bin/cli.js]: add extra space in front of `if` and remove extra space in front of `&&`

File: bin/cli.js [hunk 2]
```diff
@@ -116,7 +117,7 @@ if(previewMode) {
  * Output mode 
  */
 
-if(outputPath) {
+if(outputPath && type !== 'ts') {
```
Comment (superkhau) [inline bin/cli.js]: extra space in front of `if`

File: bin/cli.js [hunk 3]
```diff
@@ -136,8 +137,14 @@ if(outputPath) {
         console.error(err);
         process.exit();
       } else {
-        html.to(path.join(outputPath, 'index.html'));
+          html.to(path.join(outputPath, 'index.html'));
```
Comment (superkhau) [inline bin/cli.js]: remove newly added indentation

File: bin/cli.js [hunk 4]
```diff
@@ -136,8 +137,14 @@ if(outputPath) {
...
       }
     });
   });
 }
+
+if(type == 'ts'){
```
Comment (superkhau) [inline bin/cli.js]: add extra space in front of if and after )

File: bin/cli.js [hunk 5]
```diff
@@ -136,8 +137,14 @@ if(outputPath) {
...
+  ts(outputPath, previewMode);
+}
+
+
```
Comment (superkhau) [inline bin/cli.js]: remove extra empty lines

File: bin/ts.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        });
+        // Generate the docs
+        app.generateDocs(filePaths, outputPath);
+
+        if(previewMode){
+            var app = express();
+            app.use('/', express.static(process.cwd() + '/' + outputPath));
+            app.listen(3000);
+        }
+    }
```
Comment (superkhau) [inline bin/ts.js]: remove extra indentation in whole file and use 2 space indentation in entire file

File: bin/ts.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+    var fs = require('fs-extra');
+    var path = require('path');
+    var typedoc = require('typedoc');
+    var express = require('express');
+
+    module.exports = function ts(outputPath, previewMode){
+
```
Comment (superkhau) [inline bin/ts.js]: remove empty line

File: bin/ts.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+    var fs = require('fs-extra');
+    var path = require('path');
+    var typedoc = require('typedoc');
+    var express = require('express');
+
+    module.exports = function ts(outputPath, previewMode){
```
Comment (superkhau) [inline bin/ts.js]: add space after )

File: bin/ts.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+        
+        var filePaths = JSON.parse(fs.readFileSync("ts-docs.json")).content;
```
Comment (superkhau) [inline bin/ts.js]: single quotes instead of double quotes for "ts-docs.json" -> 'ts-docs.json'

File: bin/ts.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+        // Setup our TypeDoc app
```
Comment (superkhau) [inline bin/ts.js]: remove comment

File: bin/ts.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+        var app = new typedoc.Application({
+            mode:   'Modules',
```
Comment (superkhau) [inline bin/ts.js]: one space between `mode: 'Modules'`

File: bin/ts.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+            logger: 'console',
+            target: 'ES6',
+            module: 'CommonJS'
+        });
+        // Generate the docs
```
Comment (superkhau) [inline bin/ts.js]: remove comment

File: bin/ts.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+        app.generateDocs(filePaths, outputPath);
+
+        if(previewMode){
```
Comment (superkhau) [inline bin/ts.js]: space before ( and space after )

**Comment by superkhau** [comment]
@mpatsute Please review the style guide and keep the points in mind when submitting PRs -- http://loopback.io/doc/en/contrib/style-guide.html

File: bin/ts.js (hunk 6 above)
Comment (mpatsute) [inline bin/ts.js]: @superkhau - Fixed all the review comments on lint-ing and formatting.

**Comment by raymondfeng** [comment]
@mpatsute Here is my expectation per our discussion yesterday.

1. `strong-docs` should have the ability to parse both JavaScript and TypeScript files. It should be able to detect file types without explicit flags.

2. `strong-docs` should be able to parse JS and TS comments into JSON representation of such descriptors as the source of truth to generate API reference docs.

3. `strong-docs` should be able to render JS/TS descriptors into HTML pages with CSS support for consistent Look and Feel.

File: bin/ts.js (hunk 6 above)
Comment (superkhau) [inline bin/ts.js]: The indentation is still off in this file.

**Comment by hacksparrow** [comment]
Echoing @superkhau, please include tests for all the functionality you have implemented.

File: lib/doc.js [hunk 14]
```diff
@@ -67,13 +71,15 @@ var markedOptions = {
...
-function Doc(file, contents, isJS, docs) {
+function Doc(file, contents, isJS, isTS, docs) {
   var doc = this;
   this.docs = docs;
   this.file = file;
   this.isJS = isJS;
+  this.isTS = isTS;
   this.markedOptions = markedOptions;
   this.contents = contents;
+  this.file = file;
```
Comment (raymondfeng) [inline lib/doc.js]: This is a duplicate of line 77.

**Comment by raymondfeng** [comment]
@mpatsute Please format the code per our eslint style.

File: lib/docs.js [hunk 15]
```diff
@@ -157,6 +165,7 @@ Docs.prototype.parse = function (fn) {
   te.on('error', fn);
   te.on('done', function () {
     matchedFiles.forEach(function (f) {
+      console.log('processing file: ' + f);
```
Comment (bajtos) [inline lib/docs.js]: Forgotten console.log?

File: lib/doc.js [hunk 16]
```diff
@@ -67,11 +71,12 @@ var markedOptions = {
  * @constructor
  */
 
-function Doc(file, contents, isJS, docs) {
+function Doc(file, contents, isJS, isTS, docs) {
```
Comment (bajtos) [inline lib/doc.js]: boolean arguments are difficult to understand when reading the calling code.

```ts
new Doc(file, contents, false, true, docs)
```

Also the constructor has already too many arguments, adding more of them does not scale well.

I am proposing the following possible improvements:

 1. `function Doc(file, contents, flags, docs)` where `isJS = flags.isJS` and `isTS = flags.isTS`.
 2. Alternatively, replace all arguments with a single options object `function Doc(options)`.
 3. Or perhaps change `isJS` to `fileExtension` and then `isJS = fileExtension === '.js'`, etc.

File: lib/doc.js [hunk 17]
```diff
@@ -17,7 +17,11 @@ var debug = require('debug')('Doc')
   , hljs = require('highlight.js')
   , assert = require('assert')
   , string = require('underscore.string')
-  , Annotation = require('./annotation');
+  , Annotation = require('./annotation')
+  , TSConstruct = require('./tsConstruct')
+  , typedoc = require('typedoc')
+  , fs = require('fs-extra')
```
Comment (bajtos) [inline lib/doc.js]: AFAICT, this line is not used. Please remove.

File: package.json [hunk 18]
```diff
@@ -17,13 +17,16 @@
...
-    "ejs": "~0.8.4",
+    "ejs": "~2.5.6",
     "strong-task-emitter": "~0.0.5",
     "markdown": "~0.5.0",
     "highlight.js": "~7.3.0",
     "marked": "~0.2.9",
     "glob": "3.2.6",
-    "underscore.string": "~2.3.3"
+    "underscore.string": "~2.3.3",
+    "fs-extra": "^3.0.1",
```
Comment (bajtos) [inline package.json]: AFAICT, `fs-extra` is not used at all, please revert this line.

File: lib/docs.js (hunk 15 above)
Comment (mpatsute) [inline lib/docs.js]: This is intentional!

**Pull request merged**